// rg/duels_test.js
'use strict';

let Long = require('long');

let Future = require('./future');
let Client = require('./client');
let Auth = require('./auth');

function duels_test(settings)
{
	settings = settings || {};

	settings.keys = settings.keys || {};
	settings.minId = settings.minId || 1;
	settings.maxId = settings.maxId || 100;
	settings.useAdmin = (settings.useAdmin == undefined ? true : settings.useAdmin);
	settings.packetDuration = settings.packetDuration || 200;
	settings.packetsCount = settings.packetsCount || 200;
	settings.server = settings.server || {};
	settings.verbose = settings.verbose;

	let packets = settings.packets;

	let server =
	{
		host : settings.server.host,
		port : settings.server.port,
		reconnect : true,
		reconnectDuration : settings.server.reconnectDuration,
		verbose : settings.verbose,
		groupSize : settings.server.groupSize,
		deviceSize : settings.server.deviceSize
	};

	let auth = new Auth(settings.keys);
	let clients = [];

	let packetsNames = Object.keys(packets.client);

	let randomPacket = (client) =>
	{
		if (!client.connected)
			return;

		let packetName = packetsNames[~~(Math.random() * packetsNames.length)];
		let packet = packets.client[packetName];
		client.sendRandom(packet);

		Future.delay(settings.packetDuration).then(randomPacket.bind(null, client));
	};

	let duelPacket = (client) =>
	{
		if (!client.connected)
			return;

		let packetName = packetsNames[~~(Math.random() * packetsNames.length)];
		let packet = packets.client.ARENA_PLAY;
		client.sendRandom(packet);

		Future.delay(settings.packetDuration).then(randomPacket.bind(null, client));
	};

	for (let i = settings.minId; i <= settings.maxId; ++i)
	{
		let client = new Client(packets, server);
		let type = auth.randomType(settings.useAdmin);
		let netId = new Long(i);

		console.log(`create user ${type.toString()}:${netId}`);

		client.on('error', (error) => console.log(error.message));

		client.on('connect', () =>
		{
			console.log(`user ${type.toString()}:${netId} connected`);

			client.sendRaw(packets.client.HELLO)
			.then(() => client.send(packets.client.LOGIN, auth.generate(type, netId, {version: 1})))
			.catch((error) => console.error(error.stack));
		});

		client.on(packets.server.LOGIN, (packet, data) =>
		{
			if (data.status !== 0)
			{
				console.log(`client ${type}:${netId} can't login - reconnect`);
				client.disconnect();
				return;
			}

			for (let j = 0; j < settings.packetsCount; ++j)
				randomPacket(client);

			//duelPacket(client);
			client.send(packets.client.ARENA_PLAY, auth.generate(type, netId, {attacker_id: 1}))
		});

		client.connect();

		clients.push(client);
	}
}

module.exports = duels_test;
